### PR TITLE
[Feat] multi run optim args

### PR DESCRIPTION
**Pull request by Jackmin801**
> [!NOTE]
> Enables per-run configuration and correct scaling for multi-run LoRA training, and standardizes LoRA naming.
> 
> - Add per-run `optim.lr` and `model.lora` (with `name`, optional `rank`, `alpha`) to orchestrator config; trainer auto-fills/validates and auto-names when missing
> - Refactor usages from `lora_name` to `model.lora.name` (configs, CLI, orchestrator code, tests)
> - Introduce global `lora_num_tokens` and per-run `scaling_factors` tensors; update MultiLoRA layers (`multi_linear`, `multi_moe`) to consume them and apply per-token/per-adapter scaling
> - Extend `Runs` to own these tensors on-device, validate new run configs, compute/sync scaling factors, and expose per-run named params/state dicts
> - Update trainer loop to set `lora_num_tokens` (incl. CP adjustments), initialize `Runs` with device, and register LoRA rank validation/scaling hooks
> - Make optimizer/scheduler per-run aware: create optimizers/schedulers with each run’s LR; remove global LR from multi-scheduler setup
> - Minor config tweaks (add `inference.gpu_memory_utilization`, wiki/CI TOMLs)
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit f5486e029dc2ed1728d40d7fdbf403a11f11457e. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: src/prime_rl/trainer/rl/train.py [hunk 1]
```diff
@@ -91,11 +91,30 @@ def train(config: RLTrainerConfig):
...
+                orch_config.model.lora.rank = trainer_lora.rank
+            if orch_config.model.lora.rank > trainer_lora.rank:
+                return (
+                    False,
+                    f"model.lora.rank ({orch_config.model.lora.rank}) exceeds trainer max rank ({trainer_lora.rank})",
+                )
+            return True, ""
+
+        def compute_scaling(orch_config) -> float:
+            return orch_config.model.lora.alpha / orch_config.model.lora.rank
```
[the remark this answers is not in the document]
Comment (Jackmin801) [reply src/prime_rl/trainer/rl/train.py:113]: we default set it to the trainers one so shoud be fine

**Comment by Jackmin801** [comment]
@codex review

File: src/prime_rl/trainer/rl/train.py [hunk 2]
```diff
@@ -91,11 +91,30 @@ def train(config: RLTrainerConfig):
...
+
+        runs.register_config_validation_hook(validate_lora_rank)
+        runs.register_scaling_hook(compute_scaling)
```
Comment (Jackmin801) [inline src/prime_rl/trainer/rl/train.py:116]: for the record, this should probably be encapsulated somewhere else but we can refactor it in a future PR.

File: src/prime_rl/trainer/rl/train.py [hunk 3]
```diff
@@ -288,43 +307,46 @@ def train(config: RLTrainerConfig):
...
-                    # Shift down by seq idx and clip to edges of chunk
-                    lora_cu_offsets = torch.clip(lora_cu_offsets - chunk_size * cp_rank, min=0, max=chunk_size)
-                set_multilora_offsets(lora_cu_offsets)
+                    # Convert to cumsum, adjust for CP chunk, convert back to num_tokens
+                    cu_offsets = lora_num_tokens.cumsum(dim=0, dtype=torch.int32)
+                    adjusted_cu = torch.clip(cu_offsets - chunk_size * cp_rank, min=0, max=chunk_size)
+                    lora_num_tokens = torch.diff(
+                        adjusted_cu, prepend=torch.tensor([0], device=adjusted_cu.device, dtype=adjusted_cu.dtype)
+                    )
+                set_lora_num_tokens(lora_num_tokens)
```
Comment (Jackmin801) [inline src/prime_rl/trainer/rl/train.py:319]: this should have its own encapsulation as well

File: src/prime_rl/trainer/optim.py [hunk 4]
```diff
@@ -36,21 +36,27 @@ def setup_optimizer(
...
 def _create_optimizer(
-    config: OptimizerConfigType, named_params: list[tuple[str, nn.Parameter]], device_mesh: DeviceMesh
+    config: OptimizerConfigType,
+    named_params: list[tuple[str, nn.Parameter]],
+    device_mesh: DeviceMesh,
+    lr: float | None = None,
 ) -> Optimizer:
+    """Create optimizer. If lr is None, uses config.lr."""
+    if lr is None:
+        lr = config.lr
```
Comment (Jackmin801) [inline src/prime_rl/trainer/optim.py:46]: im actually not sure if we allow None right now. I guess the intended behavior was that orchestrator doesnt have to set it and if they didnt set it then we use the trainers lr. But i think theres no way to make it happen rn as a default is set in orch and it doesnt take None

File: src/prime_rl/trainer/runs.py [hunk 5]
```diff
@@ -26,7 +26,7 @@ class Progress:
 class Runs:
     """This class stores information about the runs in the system."""
 
-    def __init__(self, output_dir: Path, max_runs: int):
+    def __init__(self, output_dir: Path, max_runs: int, device: torch.device = torch.device("cpu")):
```
Comment (Jackmin801) [inline src/prime_rl/trainer/runs.py:29]: im not sure if im a fan of runs being the manager of the multilora globals. might change this in the future but lets keep it that way for now since it already works

Comment (Jackmin801) [reply src/prime_rl/trainer/runs.py:29]: also add docs for the multilora globals. theyre kind of magic rn and only someone reading the multilora module codes will figure it out and it might take them consdirable effort vs if it is just spelled out at the top of base.py or something

File: src/prime_rl/trainer/runs.py [hunk 6]
```diff
@@ -160,9 +210,21 @@ def sync_runs(self) -> None:
...
+                if self._scaling_hook is not None:
+                    self.scaling_factors[new_id] = self._scaling_hook(self.config[new_id])
+
+            # Include configs for new runs so non-master ranks have them
+            new_configs = {new_run: self.config[self.id_2_idx[new_run]] for new_run in new_runs}
+
             sync_data = {
                 "id_2_idx": self.id_2_idx,
                 "ready_to_update": self.ready_to_update,
+                "scaling_factors": self.scaling_factors.cpu(),
```
Comment (Jackmin801) [inline src/prime_rl/trainer/runs.py:226]: i believe we actually dont need to be broadcasting scaling_factors anymore because the other ranks can figure it out from the config. lets leave it for now since I think this PR works and ppl on RFT are probably dying to set the learning_rate

File: src/prime_rl/rl.py [hunk 7]
```diff
@@ -413,9 +418,17 @@ def auto_setup_lora(self):
         if self.trainer.model.lora is not None:
             if self.trainer.weight_broadcast.type == "nccl":
                 raise ValueError("NCCL weight broadcast does not support LoRA yet.")
-            if self.orchestrator.lora_name is None:
-                lora_name = f"r{self.trainer.model.lora.rank}-a{self.trainer.model.lora.alpha}"
-                self.orchestrator.lora_name = lora_name
+
+            # Ensure orchestrator has LoRA config
```
Comment (Jackmin801) [inline src/prime_rl/rl.py:422]: this comment is bad. i think it should explain that we want to let rl.py have people set it anywhere and it mirrors. i believe it doesnt mirror either rn and trainer takes precedence.

File: src/prime_rl/trainer/models/layers/lora/multi_linear.py [hunk 8]
```diff
@@ -169,7 +159,10 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
             lora_out = _run_lora_grouped_mm(lora_x, combined_lora_A, combined_lora_B, offsets)
         else:
             lora_out = _run_lora_for_loop(lora_x, combined_lora_A, combined_lora_B, offsets)
-        return (base_out + self.scaling * lora_out).view(new_shape)
+
+        # Apply per-token scaling
+        per_token_scaling = torch.repeat_interleave(self._scaling_factors, self._lora_num_tokens).unsqueeze(-1)
+        return (base_out + per_token_scaling * lora_out).view(new_shape)
```
Comment (Jackmin801) [inline src/prime_rl/trainer/models/layers/lora/multi_linear.py:165]: dont think any of this will graph break but might want to add back that test that makes sure modules dont graph break
